pulser/simulation/simulation.py

```diff
@@ -256,57 +296,214 @@ def get_hamiltonian(self, time):
             extracted from the effective sequence (determined by
             `self.sampling_rate`) at the specified time.
         """
+        # Refresh the hamiltonian
+        self._build_hamiltonian()
+
         if time > 1000 * self._times[-1]:
             raise ValueError("Provided time is larger than sequence duration.")
         if time < 0:
             raise ValueError("Provided time is negative.")
         return self._hamiltonian(time/1000)  # Creates new Qutip.Qobj
 
+    def _prepare_spam_detune(self):
+        """Choose atoms that will be badly prepared"""
+        # Tag True if atom is badly prepared
+        self.spam_detune = {qid: (np.random.uniform() < self.spam_dict['eta'])
+                            for qid in self._qid_index}
+
+    def add_noise(self, noise_type):
+        """Adds a noise model to the Simulation instance.
+            Args:
+                noise_type (str): Choose among:
+                    'doppler': Noisy doppler runs
+                    'amplitude: Noisy gaussian beam
+                    'SPAM': SPAM errors. Adds:
+                        eta: Probability of each atom to be badly prepared
+                        epsilon: false positives
+                        epsilon_prime: false negatives
+        """
+        # Check proper input:
+        noise_dict_set = {'doppler', 'amplitude', 'SPAM', 'dephasing'}
+        if noise_type not in noise_dict_set:
+            raise ValueError('Not a valid noise type')
+
+        # Add noise/error:
+        if noise_type == 'SPAM':
+            # Set SPAM parameters (experimental)
+            self.init_spam()
+            self._prepare_spam_detune()
+        if noise_type == 'doppler':
+            self.init_doppler_sigma()
+        if noise_type == 'dephasing':
+            if self.basis_name == 'digital' or self.basis_name == 'all':
+                raise ValueError("Cannot include dephasing noise in digital "
+                                 "basis.")
+            self.init_dephasing()
+        # Register added noise(s)
+        self._noise.append(noise_type)
```

Comment:
This list should not have repeated elements if one calls `self.add_noise('noise_type')` multiple times . Perhaps just throw a Warning `'Noise type already in place. Skipping instruction.'` and compare to `set(self._noise)` before updating

Reply:
Simply prints `'Noise type already in place. Skipping instruction.'` and does nothing.